README.rst
Swatch
======

``swatch`` is a parser for Adobe Swatch Exchange (ASE) files.

Copyright © 2014 Marcos A. Ojeda – http://generic.cx/


With help from documentation on ASE written by
`Carl Camera <http://iamacamera.org/default.aspx?id=109>`_,
and the ASE generator written for 
`colourlovers <https://www.colourlovers.com/>`_ by
`Chris Williams <http://www.colourlovers.com/ase.phps>`_.

``swatch.write(lst, filename)`` reads in a ``list``, as described below,
and outputs a ``.ase`` file.

``swatch.parse(filename)`` reads in an ``.ase`` file and converts it to a
``list`` of colors and palettes. Colors are ``dicts`` in the form::

    {
        'name': 'color name',
        'type': 'Process',
        'data': {
            'mode': 'RGB',
            'values': [1.0, 1.0, 1.0]
        }
    }

The values provided, and their ranges, vary between color mode. For all
color modes, the value is always a ``list`` of floats. The valid ASE modes are:

* **RGB**: three floats between [0, 1]  corresponding to RGB.
* **CMYK**: four floats between [0, 1] inclusive, corresponding to CMYK.
* **Gray**: one float between [0, 1] with 1 being white, 0 being black.
* **LAB**: three floats. The first, L, is ranged from [0, 1]. Both A and B are
  floats ranging from [-128.0, 127.0] – I believe illustrator just crops
  these to whole values, though.

Palettes (or “Color Groups,” in Adobe parlance) are also ``dicts``, but they have an
item named ``swatches`` which contains a ``list`` of colors (as above) in
the palette.::

    {
        'name': 'accent colors',
        'type': 'Color Group',
        'swatches': [
            {color}, {color}, ..., {color}
        ]
    }

Because Adobe Illustrator lets swatches exist either inside and outside
of palettes, the output of ``swatch.parse(…)`` is a ``list`` that may contain
swatches and/or palettes – i.e. ``[ swatch* palette* ]``.

Here’s an example ``dict``, with a single light grey swatch, followed by a
color group containing three more swatches::

    >>> import swatch
    >>> swatch.parse("example.ase")
    [{'data': {'mode': 'Gray', 'values': [0.75]},
      'name': 'Light Grey',
      'type': 'Process'},
     {'name': 'Accent Colors',
      'swatches': [{'data': {'mode': 'CMYK',
         'values': [0.5279774069786072,
          0.24386966228485107,
          1.0,
          0.04303044080734253]},
        'name': 'Green',
        'type': 'Process'},
       {'data': {'mode': 'CMYK',
         'values': [0.6261844635009766,
          0.5890134572982788,
          3.051804378628731e-05,
          3.051804378628731e-05]},
        'name': 'Violet Process Global',
        'type': 'Global'},
       {'data': {'mode': 'LAB', 'values': [0.6000000238418579, -35.0, -5.0]},
        'name': 'Cyan Spot (global)',
        'type': 'Spot'}],
      'type': 'Color Group'}]

Spot, Global and Process Colors
-------------------------------

Something that’s not mentioned in either Carl Camera’s or Chris William’s code
is the mention of spot, global and process colors.

There are three kinds of swatch types available to you in a ASE files: Process,
Global and Spot. Process colors are standard colors, this is the default if you
define a new color in Illustrator. As the name implies, they’re mixed from either
RGB or CMYK depending on the document color mode.

Global colors are the same thing as process colors, but they have one neat property
which is that when you update them, they are updated all throughout your artwork.
This makes them something like “color references,” which are quite useful if you’re
doing something like re-skinning some extant document.

Spot colors are implicitly global, but have the nifty property that you can create
new swatches from them based on “tints” – effectively, some screened value of that
color. The only hitch is that tints, even though they can be part of your file,
can’t be stored or exchanged as swatches. I’m on the fence as to how problematic
this is, but that’s just how it goes. Even Illustrator won’t save them out, it’s
just not supported in the app (almost certainly due to the nature of this file
format).

Caveats
-------

Finally, consider the fact that your swatches can be CMYK a mixed blessing.
While this is invariably useful if you need to import some old swatches for
print work, it will pose a challenge for accurately converting back to RGB or LAB
unless you have a copy of Illustrator handy.

If you don’t, you can always install the color profile calculator from the
(oddly named) `Little CMS <http://www.littlecms.com/>`_ package, feed it the
freely-available SWOP ICC color profile, and use the default output of sRGB
to get your colors into a somewhat usable form for the Web.

If you end up with LAB spot colors, you can always pay
`Bruce Lindbloom’s math page <http://www.brucelindbloom.com/index.html?Math.html>`_
a visit, and have a look at the relatively simple (if somewhat time-consuming)
``LAB->XYZ->RGB`` formulae he has on offer.


CHANGES.txt
v0.4.1, 02/05/2019 -- Added .gitignore, .bumpversion.cfg; reformatted README.rst (fish2000)
v0.4.0, 04/16/2014 -- More tests, docs and module writes under py2.7 and py3.4 (nsfmc)
v0.2.9, 03/23/2014 -- `swatch` can write json files, round trip from ase->json->ase (nsfmc)
v0.2.3, 12/30/2012 -- Unit tests added under swatch/tests, CHANGES file added (nsfmc)


tests/__init__.py
# encoding: utf-8
"""
swatch.tests

Solarized color palette using LAB data from
http://ethanschoonover.com/solarized

Copyright (c) 2014 Marcos A. Ojeda http://generic.cx/
All Rights Reserved
MIT Licensed, see LICENSE.TXT for details
"""
from __future__ import print_function

import pytest
import unittest

class TestSwatchParser(unittest.TestCase):
    """ Tests for parser.py """
    
    def setUp(self):
        super(TestSwatchParser, self).setUp()
        self.maxDiff = 10000
    
    def compare_with_json(self, basepath):
        import swatch, os, json
        base = os.path.join("tests", "fixtures", basepath)
        with open(base + ".json") as handle:
            ase = swatch.parse(base + ".ase")
            js = json.load(handle)
            return js, ase
    
    def test_single_swatch(self):
        js, ase = self.compare_with_json("white swatch no folder")
        self.assertEqual(js, ase, "single swatch no longer parses")
    
    def test_LAB(self):
        js, ase = self.compare_with_json("solarized")
        self.assertEqual(js, ase, "LAB test fails with solarized data")
    
    def test_empty_file(self):
        js, ase = self.compare_with_json("empty white folder")
        self.assertEqual(js, ase, "empty named folder no longer parses")
    
    def test_single_swatch_in_folder(self):
        js, ase = self.compare_with_json("single white swatch in folder")
        self.assertEqual(js, ase, "folder with one swatch no longer parses")
    
    def test_RGB(self):
        js, ase = self.compare_with_json("sampler")
        self.assertEqual(js, ase, "RGB parser test fails with sampler")
    
    def test_xterm_colors(self):
        numpy = pytest.importorskip('numpy')
        js, ase = self.compare_with_json("xterm colors")
        assert len(js[0]['swatches']) > 10
        assert len(ase[0]['swatches']) > 10
        numpy.testing.assert_almost_equal([item['data']['values'] for item in js[0]['swatches']],
                                          [item['data']['values'] for item in ase[0]['swatches']])
        js[0]['swatches'] = []
        ase[0]['swatches'] = []
        self.assertEqual(js, ase, "XTerm color dict stubs compare unequal in parser test")

class TestSwatchWriter(unittest.TestCase):
    """ Tests for writer.py """
    
    def setUp(self):
        super(TestSwatchWriter, self).setUp()
        self.maxDiff = 1000
    
    def compare_with_ase(self, basepath):
        import swatch, os, json
        base = os.path.join("tests", "fixtures", basepath)
        with open(base + ".json") as swatch_data:
            generated_ase = swatch.dumps(json.load(swatch_data))
        with open(base + ".ase", 'rb') as raw_ase:
            return raw_ase.read(), generated_ase
    
    def test_single_swatch(self):
        raw, generated = self.compare_with_ase("white swatch no folder")
        self.assertEqual(raw, generated, "single swatch has broken")
    
    def test_LAB(self):
        raw, generated = self.compare_with_ase("solarized")
        self.assertEqual(raw, generated, "LAB test fails with solarized data")
    
    def test_empty_file(self):
        raw, generated = self.compare_with_ase("empty white folder")
        self.assertEqual(raw, generated, "empty named folder no longer writes")
    
    def test_single_swatch_in_folder(self):
        raw, generated = self.compare_with_ase("single white swatch in folder")
        self.assertEqual(raw, generated, "folder with one swatch no longer writes")
    
    def test_RGB(self):
        raw, generated = self.compare_with_ase("sampler")
        self.assertEqual(raw, generated, "RGB writer test fails with sampler")
    
    def test_xterm_colors(self):
        raw, generated = self.compare_with_ase("xterm colors")
        self.assertEqual(raw, generated, "XTerm colors compare unequal in writer test")

if __name__ == '__main__':
    unittest.main()


tests/test_parser.py
# encoding: utf-8
"""
swatch.tests.test_parser

Solarized color palette using LAB data from
http://ethanschoonover.com/solarized

Copyright (c) 2019 Marcos A. Ojeda http://generic.cx/
All Rights Reserved
MIT Licensed, see LICENSE.TXT for details
"""
from __future__ import print_function

import pytest
import unittest

class TestSwatchParser(unittest.TestCase):
    """ Tests for parser.py """
    
    def setUp(self):
        super(TestSwatchParser, self).setUp()
        self.maxDiff = 10000
    
    def compare_with_json(self, basepath):
        import swatch, os, json
        base = os.path.join("tests", "fixtures", basepath)
        with open(base + ".json") as handle:
            ase = swatch.parse(base + ".ase")
            js = json.load(handle)
            return js, ase
    
    def test_single_swatch(self):
        js, ase = self.compare_with_json("white swatch no folder")
        self.assertEqual(js, ase, "single swatch no longer parses")
    
    def test_LAB(self):
        js, ase = self.compare_with_json("solarized")
        self.assertEqual(js, ase, "LAB test fails with solarized data")
    
    def test_empty_file(self):
        js, ase = self.compare_with_json("empty white folder")
        self.assertEqual(js, ase, "empty named folder no longer parses")
    
    def test_single_swatch_in_folder(self):
        js, ase = self.compare_with_json("single white swatch in folder")
        self.assertEqual(js, ase, "folder with one swatch no longer parses")
    
    def test_RGB(self):
        js, ase = self.compare_with_json("sampler")
        self.assertEqual(js, ase, "RGB parser test fails with sampler")
    
    def test_xterm_colors(self):
        numpy = pytest.importorskip('numpy')
        js, ase = self.compare_with_json("xterm colors")
        assert len(js[0]['swatches']) == 256
        assert len(ase[0]['swatches']) == 256
        numpy.testing.assert_almost_equal([item['data']['values'] for item in js[0]['swatches']],
                                          [item['data']['values'] for item in ase[0]['swatches']])
        numpy.testing.assert_allclose([item['data']['values'] for item in js[0]['swatches']],
                                      [item['data']['values'] for item in ase[0]['swatches']])
        js[0]['swatches'] = []
        ase[0]['swatches'] = []
        self.assertEqual(js, ase, "XTerm color dict stubs compare unequal in parser test")

if __name__ == '__main__':
    unittest.main()


tests/test_writer.py
# encoding: utf-8
"""
swatch.tests.test_writer

Solarized color palette using LAB data from
http://ethanschoonover.com/solarized

Copyright (c) 2019 Marcos A. Ojeda http://generic.cx/
All Rights Reserved
MIT Licensed, see LICENSE.TXT for details
"""
from __future__ import print_function

import unittest

class TestSwatchWriter(unittest.TestCase):
    """ Tests for writer.py """
    
    def setUp(self):
        super(TestSwatchWriter, self).setUp()
        self.maxDiff = 1000
    
    def compare_with_ase(self, basepath):
        import swatch, os, json
        base = os.path.join("tests", "fixtures", basepath)
        with open(base + ".json") as swatch_data:
            generated_ase = swatch.dumps(json.load(swatch_data))
        with open(base + ".ase", 'rb') as raw_ase:
            return raw_ase.read(), generated_ase
    
    def test_single_swatch(self):
        raw, generated = self.compare_with_ase("white swatch no folder")
        self.assertEqual(raw, generated, "single swatch has broken")
    
    def test_LAB(self):
        raw, generated = self.compare_with_ase("solarized")
        self.assertEqual(raw, generated, "LAB test fails with solarized data")
    
    def test_empty_file(self):
        raw, generated = self.compare_with_ase("empty white folder")
        self.assertEqual(raw, generated, "empty named folder no longer writes")
    
    def test_single_swatch_in_folder(self):
        raw, generated = self.compare_with_ase("single white swatch in folder")
        self.assertEqual(raw, generated, "folder with one swatch no longer writes")
    
    def test_RGB(self):
        raw, generated = self.compare_with_ase("sampler")
        self.assertEqual(raw, generated, "RGB writer test fails with sampler")
    
    def test_xterm_colors(self):
        raw, generated = self.compare_with_ase("xterm colors")
        self.assertEqual(raw, generated, "XTerm colors compare unequal in writer test")

if __name__ == '__main__':
    unittest.main()


swatch/parser.py
# encoding: utf-8
"""
swatch, a parser for adobe swatch exchange files
Copyright (c) 2014 Marcos A. Ojeda http://generic.cx/

With notes from:
• http://iamacamera.org/default.aspx?id=109 and
• http://www.colourlovers.com/ase.phps

All Rights Reserved
MIT Licensed, see LICENSE.TXT for details
"""
import logging
import struct
import os


def parse_chunk(handle):
    """ Generate object dicts for arbitrary chunks, until the
        filehandle `handle` has been exhausted.
    """
    chunk_type = handle.read(2)
    
    while chunk_type:
        if chunk_type == b'\x00\x01':
            # a single color
            logging.debug("[swatch.parser] parse_chunk saw single color")
            out = dict_for_chunk(handle)
            yield out
        
        elif chunk_type == b'\xC0\x01':
            # folder/palate
            logging.debug("[swatch.parser] parse_chunk saw folder")
            out = dict_for_chunk(handle)
            out['swatches'] = [x for x in colors(handle)]
            yield out
        
        elif chunk_type == b'\xC0\x02':
            # this signals the end of a folder
            logging.debug("[swatch.parser] parse_chunk saw end of folder")
            assert handle.read(4) == b'\x00\x00\x00\x00'
            pass
        
        else:
            # the file is malformed?
            logging.debug("[swatch.parser] parse_chunk got malformed ase file")
            assert chunk_type in [
                b'\xC0\x01', b'\x00\x01', b'\xC0\x02', b'\x00\x02']
            pass
        
        chunk_type = handle.read(2)


def colors(handle):
    """ Generate object dicts for all continuously found color chunks """
    chunk_type = handle.read(2)
    while chunk_type in [b'\x00\x01', b'\x00\x02']:
        out = dict_for_chunk(handle)
        yield out
        chunk_type = handle.read(2)
    handle.seek(-2, os.SEEK_CUR)

COLOR_TYPES = ('Global', 'Spot', 'Process')
COLOR_MODES = { b'RGB'  : '!fff',
                b'Gray' : '!f',
                b'CMYK' : '!ffff',
                b'LAB'  : '!fff' }

def dict_for_chunk(handle):
    """ Return a dict with decoded information for a single chunk """
    chunk_length = struct.unpack(">I", handle.read(4))[0]
    data = handle.read(chunk_length)
    
    title_length = 2 + ((struct.unpack(">H", data[:2])[0]) * 2)
    title = data[2:title_length].decode("utf-16be").strip('\0')
    color_data = data[title_length:]
    
    output = {
        'name': str(title),
        'type': 'Color Group'  # default to color group
    }
    
    if color_data:
        color_mode = struct.unpack("!4s", color_data[:4])[0].strip()
        color_values = list(struct.unpack(COLOR_MODES[color_mode], color_data[4:-2]))
        
        swatch_type_index = struct.unpack(">h", color_data[-2:])[0]
        swatch_type = COLOR_TYPES[swatch_type_index]
        
        output.update({
            'data': {
                'mode': color_mode.decode('utf-8'),
                'values': color_values
            },
            'type': str(swatch_type)
        })
    
    return output


swatch/writer.py
# encoding: utf-8
"""
swatch, a parser for adobe swatch exchange files
Copyright (c) 2014 Marcos A. Ojeda http://generic.cx/

With notes from:
• http://iamacamera.org/default.aspx?id=109 and
• http://www.colourlovers.com/ase.phps

All Rights Reserved
MIT Licensed, see LICENSE.TXT for details
"""
import struct


def chunk_count(swatch):
    """ Return the number of byte-chunks in a swatch object
        
        This function recursively walks the swatch list,
        returning 1 for a single color and 2 for each folder –
        plus 1 for each color it contains.
    """
    if type(swatch) is dict:
        if 'data' in swatch:
            return 1
        if 'swatches' in swatch:
            return 2 + len(swatch['swatches'])
    else:
        return sum(map(chunk_count, swatch))

COLOR_TYPES = ('Global', 'Spot', 'Process')
COLOR_MODES = { b'RGB'  : '!fff',
                b'Gray' : '!f',
                b'CMYK' : '!ffff',
                b'LAB'  : '!fff' }

def chunk_for_object(obj):
    """ Return an encoded byte-chunk for a color or a folder """
    chunk_type = obj.get('type')
    if chunk_type == 'Color Group':
        return chunk_for_folder(obj)
    if chunk_type in COLOR_TYPES:
        return chunk_for_color(obj)

def chunk_for_color(obj):
    """ Builds up a byte-chunk for a color
        
        The format for this is:
            b'\x00\x01' +
            Big-Endian Unsigned Int == len(bytes that follow in this block)
              • Big-Endian Unsigned Short == len(color_name)
                  …In practice, because utf-16 takes up 2 bytes per letter,
                  this’ll be 2 * (len(name) + 1) – so a color named 'foo'
                  would be 8 bytes long.
              • UTF-16BE Encoded color_name terminated with '\0'
                  …Using 'foo', this yields '\x00f\x00o\x00o\x00\x00'.
              • A 4-byte char for Color mode ('RGB ', 'Gray', 'CMYK', 'LAB ')
                  …Note the trailing spaces.
              • A variable-length number of 4-byte length floats
                  …This depends entirely on the color mode of the color.
              • A Big-Endian short int, indicating that the color chunk contains
                a global, spot, or process color:
                  global == 0,
                    spot == 1,
                 process == 2.
        
        The chunk has no terminating string – while other sites have indicated
        that the global/spot/process short is a terminator, it's actually used
        to indicate how Illustrator should deal with the color.
    """
    title = obj['name'] + '\0'
    title_length = len(title)
    chunk = struct.pack('>H', title_length)
    chunk += title.encode('utf-16be')
    
    mode = obj['data']['mode'].encode()
    values = obj['data']['values']
    color_type = obj['type']
    
    if mode in COLOR_MODES:
        padded_mode = mode.decode().ljust(4).encode()
        chunk += struct.pack('!4s', padded_mode)         # the color mode
        chunk += struct.pack(COLOR_MODES[mode], *values) # the color values
    
    if color_type in COLOR_TYPES:
        color_int = COLOR_TYPES.index(color_type)
        chunk += struct.pack('>h', color_int) # append swatch mode
    
    chunk = struct.pack('>I', len(chunk)) + chunk # prepend the chunk size
    return b'\x00\x01' + chunk # swatch color header

def chunk_for_folder(obj):
    """ Produce a byte-chunk for a folder of colors.
        
        The structure is very similar to a color's data:
        • Header:
            b'\xC0\x01' +
            Big Endian Unsigned Int == len(Bytes in the Header Block)
              note _only_ the header, this doesn't include the length of color data
              • Big Endian Unsigned Short == len(Folder Name + '\0')
                  Note that «Folder Name» is assumed to be utf-16be,
                  so this will always be an even number
              • Folder Name + '\0', encoded UTF-16BE
        • Body:
            Chunks for each color, see ``chunk_for_color``
        • folder terminator
            b'\xC0\x02' +
            b'\x00\x00\x00\x00'
        
        Perhaps the four null bytes represent something; I'm pretty sure
        they're just a terminating string – but there's something nice about
        how the b'\xC0\x02' matches with the folder's header.
    """
    title = obj['name'] + '\0'
    title_length = len(title)
    chunk_body = struct.pack('>H', title_length) # title length
    chunk_body += title.encode('utf-16be') # title

    chunk_head = b'\xC0\x01' # folder header
    chunk_head += struct.pack('>I', len(chunk_body))
    # precede entire chunk by folder header and size of folder
    chunk = chunk_head + chunk_body

    chunk += b''.join([chunk_for_color(c) for c in obj['swatches']])

    chunk += b'\xC0\x02' # folder terminator chunk
    chunk += b'\x00\x00\x00\x00' # folder terminator
    return chunk


utils/dl-xterm-color-list.py
#!/usr/bin/env python
# encoding: utf-8
from __future__ import print_function

import json
import math
import os
import requests
import shutil
import swatch
import tempfile

from bs4 import BeautifulSoup

try:
    string_type = basestring
except NameError:
    string_type = str

def hex2rgb(h):
    """ Convert a hex string or number to an RGB triple """
    # q.v. https://git.io/fh9E2
    if isinstance(h, string_type):
        return hex2rgb(int(h[1:] if h.startswith('#') else h, 16))
    return (h >> 16) & 0xff, (h >> 8) & 0xff, h & 0xff

def compand(v):
    """ Compand a linearized value to an sRGB byte value """
    # q.v. http://www.brucelindbloom.com/index.html?Math.html
    V = (v <= 0.0031308) and (v * 12.92) or ((1.055 * math.pow(v, 1 / 2.4)) - 0.055)
    return V * 255.0

def uncompand(A):
    """ Uncompand an sRGB byte value to a linearized value """
    # q.v. http://www.brucelindbloom.com/index.html?Eqn_RGB_to_XYZ.html
    V = A / 255.0
    return (V <= 0.04045) and (V / 12.92) or math.pow(((V + 0.055) / 1.055), 2.4)

def tds_to_dictionary(tds):
    """ Package the scraped XTerm color data as a dictionary """
    # q.v. https://jonasjacek.github.io/colors/
    out = dict(xterm_number=int(tds[1].contents[0]),
               name=tds[2].contents[0].strip(),
               hex=tds[3].contents[0].strip(),
               rgb_string=tds[4].contents[0].strip(),
               hsl_string=tds[5].contents[0].strip(),
               rgb=hex2rgb(tds[3].contents[0].strip()))
    return out

def scraped_to_ase(scraped_color):
    """ Convert a scraped XTerm color dictionary to an ASE dictionary """
    # q.v https://github.com/fish2000/swatch/blob/master/README.rst
    assert 'name' in scraped_color
    assert 'rgb' in scraped_color
    values = [uncompand(A) for A in scraped_color['rgb']]
    out = dict(name=scraped_color['name'],
               type='Process',
               data=dict(mode='RGB', values=values))
    return out

def scrape(url, fixtures_dir):
    response = requests.get(url)
    soup = BeautifulSoup(response.text, features='lxml')
    scraped = []
    output = [{ 'name'      : 'XTerm Colors',
                'type'      : 'Color Group',
                'swatches'  : [] }]
    
    # get all table rows:
    alltrs = soup.findAll(name='tr')
    
    # first row is the header, skip it
    trs = alltrs[1:]
    assert len(trs) == 256
    
    for tr in trs:
        tds = tuple(tr.children)
        assert len(tds) == 6
        color_dict = tds_to_dictionary(tds)
        scraped.append(color_dict)
    assert len(scraped) == 256
    
    # print(json.dumps(scraped, indent=4))
    
    for infodict in scraped:
        output[0]['swatches'].append(scraped_to_ase(infodict))
    
    print(json.dumps(output, indent=4))
    
    mode = "w"
    buffer_size = -1
    descriptor, filename = tempfile.mkstemp(suffix=".json")
    with os.fdopen(descriptor, mode, buffer_size) as handle:
        json.dump(output, handle)
        handle.flush()
    
    shutil.copy2(os.path.join(tempfile.gettempdir(), filename),
                 os.path.join(fixtures_dir, 'xterm colors.json'))
    
    mode = "wb"
    descriptor, filename = tempfile.mkstemp(suffix=".ase")
    with os.fdopen(descriptor, mode, buffer_size) as handle:
        with open(os.path.join(fixtures_dir, 'xterm colors.json'), "r") as jsonhandle:
            swatch.dump(json.load(jsonhandle), handle)
            handle.flush()
    
    shutil.copy2(os.path.join(tempfile.gettempdir(), filename),
                 os.path.join(fixtures_dir, 'xterm colors.ase'))

URL = 'https://jonasjacek.github.io/colors/'
FIXTURES = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'tests', 'fixtures')

if __name__ == '__main__':
    scrape(URL, FIXTURES)


swatch/__init__.py
# encoding: utf-8
"""
swatch, a parser for adobe swatch exchange files
Copyright (c) 2014 Marcos A. Ojeda http://generic.cx/

With notes from
http://iamacamera.org/default.aspx?id=109 by Carl Camera and
http://www.colourlovers.com/ase.phps by Chris Williams

All Rights Reserved
MIT Licensed, see LICENSE.TXT for details
"""


__title__ = 'swatch'
__version__ = '0.4.1'
__author__ = 'Marcos Ojeda'
__license__ = 'MIT'
__copyright__ = '© 2014 Marcos A. Ojeda'


from . import parser
from . import writer
import struct

HEADER = b'ASEF'
V_MAJOR = 1
V_MINOR = 0

def parse(filename):
    """ Parses a ``.ase`` file and returns a list of colors and color groups
        
        `swatch.parse(…)` reads in an ``.ase`` file and converts it to a list
        of colors and palettes. Colors are ``dict``s of the form
        
        ```json
        {
            'name': 'color name',
            'type': 'Process',
            'data': {
                'mode': 'RGB',
                'values': [1.0, 1.0, 1.0]
            }
        }
        ```
        
        The values provided vary between color mode. For all color modes, the
        value is always a ``list`` of floats. Valid ASE color modes are:
        
        RGB: three floats between [0, 1] corresponding to RGB.
        CMYK: four floats between [0, 1] inclusive, corresponding to CMYK.
        Gray: one float between [0, 1] with 1 being white, 0 being black.
        LAB: three floats. The first, L, is ranged from [0, 1]. Both A and B are
        floats ranging from [-128.0, 127.0]. I believe Illustrator just crops
        these to whole values, though.
        
        Palettes (née Color Groups in Adobe parlance) are also ``dicts``, but they have an
        item named “swatches” which contains a ``list`` of colors contained within
        the palette.
        
        ```json
        {
            'name': 'accent colors',
            'type': 'Color Group',
            'swatches': [
                {color}, {color}, ..., {color}
            ]
        }
        ```
        
        Because Adobe Illustrator lets swatches exist either inside and outside
        of palettes, the output of swatch.parse is a list that may contain
        swatches and palettes, i.e. `[ swatch* palette* ]`
        
        Here's an example with a light grey swatch followed by a color group,
        containing three colors:
        
            >>> import swatch
            >>> swatch.parse("example.ase")
            [{'data': {'mode': 'Gray', 'values': [0.75]},
              'name': 'Light Grey',
              'type': 'Process'},
             {'name': 'Accent Colors',
              'swatches': [{'data': {'mode': 'CMYK',
                 'values': [0.5279774069786072,
                  0.24386966228485107,
                  1.0,
                  0.04303044080734253]},
                'name': 'Green',
                'type': 'Process'},
               {'data': {'mode': 'CMYK',
                 'values': [0.6261844635009766,
                  0.5890134572982788,
                  3.051804378628731e-05,
                  3.051804378628731e-05]},
                'name': 'Violet Process Global',
                'type': 'Global'},
               {'data': {'mode': 'LAB', 'values': [0.6000000238418579, -35.0, -5.0]},
                'name': 'Cyan Spot (global)',
                'type': 'Spot'}],
              'type': 'Color Group'}]
    """
    with open(filename, "rb") as data:
        header, v_major, v_minor, chunk_count = struct.unpack("!4sHHI", data.read(12))
        
        assert header == HEADER
        assert (v_major, v_minor) == (V_MAJOR, V_MINOR)
        
        return [c for c in parser.parse_chunk(data)]

def dumps(obj):
    """ Converts a swatch to bytes, suitable for writing """
    chunk_count = writer.chunk_count(obj)
    head = struct.pack('!4sHHI', HEADER, V_MAJOR, V_MINOR, chunk_count)
    body = b''.join([writer.chunk_for_object(c) for c in obj])
    return head + body

def dump(obj, handle):
    """ Write a swatch to a python file object """
    handle.write(dumps(obj))

def write(obj, filename):
    """ Write a swatch object to the filename specified
        
        If `filename` exists, it will be overwritten.
        
        `obj` *must* be a list of swatches and palettes, as follows:
        
        ```
            [ swatch* palette* ]
        ```
        
        The best source for descriptions of each of these is to be found
        in the `parser` documentation.
    """
    with open(filename, 'wb') as handle:
        dump(obj, handle)
